presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -317,7 +317,7 @@ private void processNewSources()
                 }
 
                 // if current operator is finished...
-                if (current.isFinished()) {
+                if (current.isFinished() && !next.isFinished() && isBlocked(next).isDone()) {
```

Comment:
Currently, the design of the interface between Drivers and Operators is that Operator is required to deal with it's state machine by itself.  The blocked future is simply treated as a suggestion from the Operator and the Driver is (and does) free to choose to ignore this suggestion.  If there is a operator that can is having problems with this, then it is a bug in the operator. 

We can talk about changing this design, but that is a pretty big undertaking (and in my mind without much benefit).  As for this specific change, I don't see why we would want to make this.  To me this is just masking over a problem in some other operator.